src/tpm2-crypto.c

```diff
@@ -1,9 +1,130 @@
 #include <stdio.h>
+#include <string.h>
+#include <ctype.h>      // isprint, for hexdump
 #include <tss2/tss2_esys.h>
 #include <tss2/tss2_mu.h>
 #include <tss2/tss2_tpm2_types.h>
 #include <tss2/tss2_esys.h>
 #include <tss2/tss2_tctildr.h>
+#include <qcbor/qcbor_encode.h>
+
+static ESYS_CONTEXT          *esys_ctx;
+static TSS2_TCTI_CONTEXT     *tcti_ctx;
+static TPM2B_PUBLIC          *keyPublic;
+static TPM2B_PRIVATE         *keyPrivate;
+
+void hexdump(const void *memory, size_t length);
+void hexdump(const void *memory, size_t length)
+{
+	size_t i, j;
+	uint8_t *line;
+	int all_zero = 0;
+	int all_one = 0;
+	size_t num_bytes;
+
+	for (i = 0; i < length; i += 16) {
+		num_bytes = length - i;
+        if (num_bytes > 16) num_bytes = 16;
+		line = ((uint8_t *)memory) + i;
+
+		all_zero++;
+		all_one++;
+		for (j = 0; j < num_bytes; j++) {
+			if (line[j] != 0) {
+				all_zero = 0;
+				break;
+			}
+		}
+
+		for (j = 0; j < num_bytes; j++) {
+			if (line[j] != 0xff) {
+				all_one = 0;
+				break;
+			}
+		}
+
+		if ((all_zero < 2) && (all_one < 2)) {
+			printf("%#8lx:", i);
+			for (j = 0; j < num_bytes; j++)
+				printf(" %02x", line[j]);
+			for (; j < 16; j++)
+				printf("   ");
+			printf("  ");
+			for (j = 0; j < num_bytes; j++)
+				printf("%c", isprint(line[j]) ? line[j] : '.');
+			printf("\n");
+		} else if ((all_zero == 2) || (all_one == 2)) {
+			printf("...\n");
+		}
+	}
+}
+
+static UsefulBuf _encode_aik(UsefulBuf buf)
+{
+    /* From TCG Trusted Platform Module Library Part 2: Structures rev 01.59:
+     *   A TPM (...) supporting RSA shall support two primes and an exponent of
+     *   zero. An exponent of zero indicates that the exponent is the default of
+     *   2^16 + 1. Support for other values is optional.
+     */
+    /* TODO: is endianness correct? */
+    uint32_t exp_val = keyPublic->publicArea.parameters.rsaDetail.exponent;
+    if (exp_val == 0)
+        exp_val = 0x00010001;
+
+    UsefulBufC exponent = {&exp_val, sizeof(exp_val)};
+    TPM2B_PUBLIC_KEY_RSA *unique = (TPM2B_PUBLIC_KEY_RSA *)
+                                   &keyPublic->publicArea.unique;
+    UsefulBufC modulus = {&unique->buffer, unique->size};
+    /* TODO */
+    static const uint8_t ek_cert_buf[] = {0,1,2,3,4,5,6,7,8,9};
+    UsefulBufC ek_cert = {&ek_cert_buf, sizeof(ek_cert_buf)};
+
+   /* Set up the encoding context with the output buffer */
+    QCBOREncodeContext ctx;
+    QCBOREncode_Init(&ctx, buf);
+
+    /* Proceed to output all the items, letting the internal error
+     * tracking do its work */
+    QCBOREncode_OpenMap(&ctx);
+        QCBOREncode_AddInt64ToMap(&ctx, "type", 1);     /* Assume RSA */
+        QCBOREncode_OpenMapInMap(&ctx, "key");
+            QCBOREncode_AddBytesToMap(&ctx, "n", modulus);
+            QCBOREncode_AddBytesToMap(&ctx, "e", exponent);
```

Comment:
Fobnail expects exponent encoded as u32, not byte array
```suggestion
            QCBOREncode_AddUInt64ToMap(&ctx, "e", exp_val);
```